content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+* The API consistently shows which methods and functions can throw exceptions, by the consistent use of the `noexcept` keyword. This is also part of the documentation.
+* The API documentation explicitly states which types of exception can be thrown, as to allow the user of the API to determine which exceptions to handle.
+* The API introduces the [`BPrivate::Network::BError`](https://www.haiku-os.org/docs/api/classBPrivate_1_1Network_1_1BError.html) abstract base class, which defines a standard API for all derived exceptions. Part of this base class is the `origin` property, which holds the location of where the exception occured in order to aid debugging.
+* Next to the exceptions derived from `BError`, users of the API will also find exceptions of the `std::bad_alloc` type for when allocations fail due to the fact that the system is low on memory.
+
+In the internal implementation of the kit, exceptions are also used to handle error conditions and other exceptional states. Inside the code that runs in the threads that execute HTTP requests, the [`BNetworkRequestError`](https://www.haiku-os.org/docs/api/classBPrivate_1_1Network_1_1BNetworkRequestError.html) is thrown when there are network or protocol errors. For example, when there is a parsing error of an incoming response, the [exception is raised and caught](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpSession.cpp?h=r1beta4#n441), and then [saved as part of the `HttpResult` object](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpResultPrivate.h?h=r1beta4#n92) which is then [re-thrown](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpResult.cpp?h=r1beta4#n217) whenever the user tries to access data on that object.
+
+The discussion on whether or not exceptions should be used are mostly ideological and fundamental in nature. This preview allows developers to get some hands-on first hand experience of using it in the context of Haiku development, in order to better form their arguments on whether or not the future of the Haiku API is in way of exceptions, or through alternative to `status_t` return values like the [`std::expected`](https://en.cppreference.com/w/cpp/utility/expected) type, which is going to be part of C++23. Or a mixture; there might be good arguments to apply both strategies.
+
+### Thread-safe data access with Exclusive Borrows
+
+One of the biggest challenges in software development is about writing code that handles reading and writing data safely over multiple threads. The main rule is that objects should be owned, read and written by one single thread. As soon as data is shared with more than one thread, there should be mechanisms to synchronize read and write access to the data. The most common methods of synchronization are atomic data structures and semaphores/mutexes. You **need** to employ these synchronization mechanisms, otherwise you will introduce nasty bugs in your code.
```

Comment:
Generally I think this article needs an introduction with a bit of backstory about the previous attempts at libnetservices and their shortcoming.

In particular for the threads and synchronization, I think we tried 3 different ways:
- With user-provided callbacks called from the netservices request threads, letting the user of the API figure out how to do their locking (simpler design, more flexible, but maybe too flexible). People tend to not be so good at locking and maybe it would be better to handle that inside the library
- With fully asynchronous events based on BMessage: easy integration with the existing APIs, but the fully asynchronous system makes it impossible to do some things (such as cancel a request early if the HTTP headers are not what was expected, for example to not display the content of an HTTP redirect). Also BMessage/BLooper proved to have performance problems when using it in this way
- With a mix of asynchronous callbacks where needed, and a generic BDataIO based interface for the data: solved the performance problems from the above, but not the asychronous problem

I'm not sure how the borrowing concept fits into this, my understanding in Rust is that borrowing is related to memory ownership and a way to avoid reference counting and garbage collectors by ensuring ownership at compile time.

In a threading concept, what you describe below is more related to mutexes.

Also, I don't know if that's a good way to go here, I think we are in a relatively simple producer-consumer case, and in that situation, it's possible to do things in many other ways, with atomics and even in almost-lockless ways (no locking needed as long as the buffer is not empty or full).

So, in the end, all we have here is a proper C++ design for a mutex that actually blocks the protected resource from being used, with compile time checks? That's nice in many cases, but not what I would have gone with in this specific one. I don't know if there are future plans here and an attempt to not overdesign things (which is good), if there are, this maybe deserves a mention here.

Reply:
To the first point: I will add a bit more details about previous iterations. As for the rest, it feels like I should probably write another article on the particular problem `BExclusiveBorrow<T>` is trying to solve, to work through the alternatives and share why I felt those were not the best solution for this problem.